packages/puppeteer-core/src/common/Target.ts

```diff
@@ -271,15 +223,118 @@ export class Target {
    */
   _targetInfoChanged(targetInfo: Protocol.Target.TargetInfo): void {
     this.#targetInfo = targetInfo;
+    this._checkIfInitialized();
+  }
 
-    if (
-      !this._isInitialized &&
-      (!this._isPageTargetCallback(this.#targetInfo) ||
-        this.#targetInfo.url !== '')
-    ) {
+  /**
+   * @internal
+   */
+  protected _initialize(): void {
+    // TODO: refactor to deferred promises.
+    this._isInitialized = true;
+    if (this._isInitialized) {
+      this._initializedCallback(true);
+    }
+  }
+
+  /**
+   * @internal
+   */
+  protected _checkIfInitialized(): void {
+    if (!this._isInitialized) {
       this._isInitialized = true;
       this._initializedCallback(true);
       return;
     }
   }
+
+  /**
+   * If the target is not of type `"page"`, `"webview"` or `"background_page"`,
+   * returns `null`.
+   */
+  async page(): Promise<Page | null> {
+    return null;
+  }
+}
+
+/**
+ * @internal
+ */
+export class PageTarget extends Target {
+  #defaultViewport?: Viewport;
+  protected pagePromise?: Promise<Page>;
+  #screenshotTaskQueue: TaskQueue;
+  #ignoreHTTPSErrors: boolean;
+
+  /**
+   * @internal
+   */
+  constructor(
+    targetInfo: Protocol.Target.TargetInfo,
+    session: CDPSession | undefined,
+    browserContext: BrowserContext,
+    targetManager: TargetManager,
+    sessionFactory: (isAutoAttachEmulated: boolean) => Promise<CDPSession>,
+    ignoreHTTPSErrors: boolean,
+    defaultViewport: Viewport | null,
+    screenshotTaskQueue: TaskQueue
+  ) {
+    super(targetInfo, session, browserContext, targetManager, sessionFactory);
+    this.#ignoreHTTPSErrors = ignoreHTTPSErrors;
+    this.#defaultViewport = defaultViewport ?? undefined;
+    this.#screenshotTaskQueue = screenshotTaskQueue;
+  }
+
+  protected override _initialize(): void {
+    this._initializedPromise
+      .then(async success => {
+        if (!success) {
+          return false;
+        }
+        const opener = this.opener();
+        if (!(opener instanceof PageTarget)) {
+          return true;
+        }
+        if (!opener || !opener.pagePromise || this.type() !== 'page') {
+          return true;
+        }
+        const openerPage = await opener.pagePromise;
+        if (!openerPage.listenerCount(PageEmittedEvents.Popup)) {
+          return true;
+        }
+        const popupPage = await this.page();
+        openerPage.emit(PageEmittedEvents.Popup, popupPage);
+        return true;
+      })
+      .catch(debugError);
+    this._checkIfInitialized();
+  }
+
+  override async page(): Promise<Page | null> {
+    if (!this.pagePromise) {
+      const session = this._session();
```

Comment:
Can't we await 
```suggestion
      const session = this._session() ?? (await this._sessionFactory()(true));
```
then set `this.pagePromise = CDPPage._create` , I think it would easier to read.

Reply:
I don't think so because we need pagePromise to be created atomically.